Feature: add an utility to scan cache

Wauplin opened the pull request:
Following issue https://github.com/huggingface/huggingface_hub/issues/972.

This is a draft of what the scan of the cache directory could look like. At the moment there is only a `scan_cache` method that returns structured information about the cache.

I think the most important question is what information do we want in `HFCacheInfo`, `CachedRepoInfo`, `CachedRevisionInfo` and `CachedFileInfo` objects. In theory, the current content would be enough to make the utils that @sgugger described in the issue. Another information that I have not yet scanned is "last_updated" timestamp. I guess that could be useful to "prune" a repo to keep only the latest downloaded version (to confirm...).


Any suggestion / feedback / comment is very welcomed as this is a draft (ping @osanseviero @sgugger @LysandreJik @mariosasko who showed interest in this feature).

---
~~I also made a quick and dirty script (`scan.py`) to show what a CLI command could display:~~

**EDIT:** refer to the updated CLI example in https://github.com/huggingface/huggingface_hub/pull/990#issuecomment-1227166728

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_990). All of your documentation changes will be reflected on that endpoint.

Wauplin commented:
**Side note:** this only scans the "new" cache layout and makes some assumptions on the expected structure. I am not yet sure all weird cases are handled. Especially:
1. Is it possible to have a blob that is never referenced (no symlink to it ?)
2. Is it possible that a symlink refers to a non-existing blob ?
3. Is it possible that a blob is not fully downloaded ?
4. Is there a cache lock that I should be aware of ? If someone is scanning + downloading at the same time, results might not be accurate.

sgugger commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+
+    A revision can be either referenced by 1 or more `refs` or be "detached" (no refs).
+
+    /!\ `size_on_disk` is not necesarily the sum of all file sizes because of possible
+        duplicated files. Also only blobs are taken into account, not the (neglectable)
+        size of folders and symlinks.
```
```suggestion
    <Tip warning={true}>

    `size_on_disk` is not necesarily the sum of all file sizes because of possible
    duplicated files. Also only blobs are taken into account, not the (neglectable)
    size of folders and symlinks.

    </Tip>
```
If you want this to be rendered as a warning block in the doc.

Wauplin replied:
Thanks for the review and the tips ! 👍 

I'll make the changes when I come back to this PR. I first wait for other feedback and then will properly write the CLI utils + tests + proper docstring + the remaining stuff.

sgugger commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+class CachedRepoInfo:
+    r"""Information about a cached repository (dataset or model).
+
+    /!\ `size_on_disk` is not necesarily the sum of all revisions sizes because of
```
Same here.

sgugger commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+class HFCacheInfo:
+    r"""Information about the entire cache repository.
+
+    /!\ Here `size_on_disk` is equal to the sum of all repo sizes (only blobs).
```
And there.

sgugger commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+
+    datasets: List[CachedRepoInfo] = []
+    models: List[CachedRepoInfo] = []
```
Type annotations in signatures are fine, but in code I often find they make it harder to read than really illuminate things (especially when variables are aptly named like here).

julien-c replied:
personally i like them (but i like types in general unlike @sgugger 🙈 )

Wauplin replied:
I'll keep them. Typing doesn't hurt 🙄

sgugger commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+        raise CorruptedCacheException("Repo type must be `dataset` or `model` !")
+    repo_type = repo_type[:-1]  # "models" -> "model"
+
+    blob_sizes: Dict[Path, int] = {}  # Key is blob_path, value is blob size (in bytes)
```
Same here, the comment gives all the info needed, making the type hint irrelevant.

sgugger commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+    cached_revisions_by_hash: Dict[str, CachedRevisionInfo] = {}
+    for revision_path in snapshots_path.iterdir():
+        if revision_path.is_file():
+            raise CorruptedCacheException("Snapshots folder corrupted: found a file  !")
```
Here and everywhere below, maybe add useful info to the exceptions? Like the `revision_path` value here.

sgugger reviewed (approved):
LGTM! Thanks a lot for diving into this. Left a couple of comments but I don't feel strongly about anything.

osanseviero reviewed (commented):
The approach looks good to me! Looking forward the final PR :)

Wauplin commented:
@osanseviero If you have time could you try to checkout to this branch and run `python scan.py` and copy-paste the output ? I'd like to know if you get any errors. I made some tests with my local cache but it's a very recent and tiny cache. Would be good to get feedback from someone that is more likely to have more discrepancies in its cache.

If too complicated, I will update the PR with a proper CLI when I have time.

osanseviero commented:
I got the following error

```
python scan.py
Error while scanning /home/osanseviero/.cache/huggingface/hub/models--bert-base-cased: Snapshots dir doesn't exist !
Datasets:

Models:
  osanseviero--wine-quality 268.0KiB (5 files)
    /home/osanseviero/.cache/huggingface/hub/models--osanseviero--wine-quality
    Revisions:
        974ee1d10a13d2cb802c5b9c26b504a9c58a09af: main
          268.0KiB (5 files)

```

Wauplin commented:
@osanseviero Thanks for trying it out ! 

I am not exactly sure what happened here (surely a cache folder without `snapshots` folder, but why ?).
I will update the script to make it robust to errors (just log and ignore them) + add better error messages. I'll ping you when I have a better version to test.

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+    /!\ Here `size_on_disk` is equal to the sum of all repo sizes (only blobs).
+    """
+
+    datasets: List[CachedRepoInfo]
+    models: List[CachedRepoInfo]
```
i wouldn't split by types like this, and would just store `repos` (and users can filter for just e.g. models if they want to)

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+
+    datasets: List[CachedRepoInfo] = []
```
same comment here, and by the way you're missing the spaces:)

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+        raise CorruptedCacheException("Not a valid HuggingFace cache directory !")
+
+    repo_type, repo_id = repo_path.name.split("--", maxsplit=1)
+    if repo_type not in {"datasets", "models"}:
```
also spaces=)

Wauplin replied:
ok thanks, I was wondering 👍

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+            print(f"Error while scanning {repo_path}: {e}")
+
+    return HFCacheInfo(
+        datasets=sorted(datasets, key=lambda repo: repo.repo_path),
+        models=sorted(models, key=lambda repo: repo.repo_path),
```
the reason you're sorting is for determinism of output across operating systems?

Wauplin replied:
Hmm yes and also to be 100% sure that the output is in alphabetical order. I'm thinking about a user that wants to see a specific model in a long list, it will be easier if it's somehow sorted. 
(and also it really cost nothing to sort a list of max ~1000 items)

Wauplin replied:
(I would expect `Path.glob("*")` to be already sorted but it's not guaranteed)

Wauplin replied:
Actually I changed the dataclasses to "frozen" so that I can store them in sets instead of lists. Makes much more sense and no need to sort them in `scan_cache` helper.

Then only when I print the result in the CLI, I sort the output by path but not before.

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+    # Scan over `refs` directory
+    if refs_path.exists():
+        # Example of `refs` directory
+        # ── refs
+        #     ├── main
+        #     └── refs
+        #         └── pr
+        #             └── 1
```
lol, I actually hadn't realized this file structure could be nested given that we allow `/` inside branches or tag names, @LysandreJik 

I'm not 100% sure this is robust though (on Windows for instance...) but i don't think it's a big deal either way

Wauplin replied:
Yep, discovered it quite randomly while working on that. I'd say it's robust enough because we would have known by now (🙄)

Wauplin replied:
The issue is if someone copy-pastes a cache folder from a Windows machine to a Unix-based one or the opposite.
We can always fix this in a separate PR but it seems a quite niche problem no ?

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+    return CachedRepoInfo(
+        repo_id=repo_id,
+        repo_type=repo_type,
+        revisions=sorted(
+            cached_revisions_by_hash.values(), key=lambda revision: revision.commit_hash
+        ),
```
kinda same question about the sorting here (any reason you sort?)

Wauplin replied:
see above https://github.com/huggingface/huggingface_hub/pull/990#discussion_r954603419

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -0,0 +1,246 @@
+    )
+
+
+def _format_size(num: int, suffix: str = "B") -> str:
+    """Format size in bytes into a human-readable string.
+
+    Taken from https://stackoverflow.com/a/1094933
+    """
+    for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
+        if abs(num) < 1024.0:
+            return f"{num:3.1f}{unit}{suffix}"
+        num /= 1024.0
+    return f"{num:.1f}Yi{suffix}"
```
personally i think now most people stick to SI units ie. powers of 10 not of 2

aka `MB` not `MiB`

It also makes the code simpler :)

WDYT?

julien-c replied:
(@Pierrci for this too)

julien-c replied:
also consistent with the Hub where we display GB not GiB: https://huggingface.co/gpt2/tree/main

Pierrci replied:
yes, +1 on consistency and using GB instead of GiB :)

julien-c replied:
my OS (which is the superior OS™️) also displays GB... I think at this point `GiB` are for dinosaurs =)

Wauplin replied:
Yep, will update this. Haven't thought too much about it to be honest 😁

julien-c reviewed (commented):
left a few comments, looks generally good to me

Another question/comment: maybe the `CachedRepoInfo` should also expose a dict of refs, or maybe a helper to get a `CachedRevisionInfo` from a ref 

About the pruning, IMO we can keep it for later and/or leave it as the user's responsibility. It will be hard to find a single strategy that works well to decide what to prune, so it should maybe be up to the user to implement. For instance, you mention `lastModified` (of a file on disk?) but i have an old snapshot of gpt2 I use all the time... Or you have a lastFileAccess file stat on some OSes but not sure it's robust either...

Wauplin commented:
> About the pruning, IMO we can keep it for later and/or leave it as the user's responsibility. It will be hard to find a single strategy that works well to decide what to prune, so it should maybe be up to the user to implement. For instance, you mention lastModified (of a file on disk?) but i have an old snapshot of gpt2 I use all the time... Or you have a lastFileAccess file stat on some OSes but not sure it's robust either...

@julien-c Yep I talked about `lastUpdated` date because it's the only robust-enough timestamp that I thought of. I saw it more as an information to display to the user rather than a value to decide if we want to delete the file or not.

In general pruning/cache management will be done separately. My opinion is that we can have an helper to delete a specific `revision` from a specific `repo_id`. The helper deletes the blob files that are symlink-ed only to this `revision` so that other revisions are still valid. In addition it also delete and `refs` that points to this `revision`. This helper is generic-enough so that we don't take any decisions on what to delete and the user can define itself about its pruning stategy.

```py
def delete_cached_revision(repo_id: str, revision: str) -> None:
    ...
```

Wauplin commented:
> Another question/comment: maybe the CachedRepoInfo should also expose a dict of refs, or maybe a helper to get a CachedRevisionInfo from a ref

Good idea, I've added a `CachedRepoInfo.refs` property that computes the mapping "ref -> revision info"

julien-c commented on src/huggingface_hub/utils/cache_explorer.py:
```diff
@@ -239,8 +253,8 @@ def _format_size(num: int, suffix: str = "B") -> str:
             return f"{num:3.1f}{unit}{suffix}"
-        num /= 1024.0
+        num /= 1000.0
     return f"{num:.1f}Yi{suffix}"
```
i think you need to remove the `i` here?

Wauplin replied:
Indeed, thanks for the catch 
Would have displayed a `YiB` in case someone has a file of 10^24 bytes 😁

julien-c replied:
haha yeah that would be one large ML model

Wauplin commented:
I took the feedback, made some changes and added a proper command to the existing `huggingface-cli` tool. I also changed printed output to have something more tabular and grep-able. This is most likely the final version of this tool for this PR. If you have any feedback, please tell me.

In the meantime I will write documentation and docstrings for this tool.

```
➜ huggingface-cli scan-cache --help
usage: huggingface-cli <command> [<args>] scan-cache [-h] [--dir DIR] [-v]

options:
  -h, --help     show this help message and exit
  --dir DIR      cache directory to scan (optional). Default to the default HuggingFace cache.
  -v, --verbose  show a more verbose output
```
```
➜ huggingface-cli scan-cache       
REPO ID                     REPO TYPE SIZE ON DISK NB FILES REFS                LOCAL PATH                                                                
--------------------------- --------- ------------ -------- ------------------- ------------------------------------------------------------------------- 
glue                        dataset         116.3K       15 2.4.0, main, 1.17.0 /Users/lucain/.cache/huggingface/hub/datasets--glue                       
google/fleurs               dataset          64.9M        6 refs/pr/1, main     /Users/lucain/.cache/huggingface/hub/datasets--google--fleurs             
Jean-Baptiste/camembert-ner model           441.0M        7 main                /Users/lucain/.cache/huggingface/hub/models--Jean-Baptiste--camembert-ner 
bert-base-cased             model             1.9G       13 main                /Users/lucain/.cache/huggingface/hub/models--bert-base-cased              
t5-base                     model            10.1K        3 main                /Users/lucain/.cache/huggingface/hub/models--t5-base                      
t5-small                    model           970.7M       11 refs/pr/1, main     /Users/lucain/.cache/huggingface/hub/models--t5-small                     

Done in 0.0s. Scanned 6 repo(s) for a total of 3.4G.
Got 1 error(s) while scanning. Use -vvv to print details.
```
```
➜ huggingface-cli scan-cache -v
REPO ID                     REPO TYPE REVISION                                 SIZE ON DISK NB FILES REFS        LOCAL PATH                                                                                                                   
--------------------------- --------- ---------------------------------------- ------------ -------- ----------- ---------------------------------------------------------------------------------------------------------------------------- 
glue                        dataset   9338f7b671827df886678df2bdd7cc7b4f36dffd        97.7K       14 main, 2.4.0 /Users/lucain/.cache/huggingface/hub/datasets--glue/snapshots/9338f7b671827df886678df2bdd7cc7b4f36dffd                       
glue                        dataset   f021ae41c879fcabcf823648ec685e3fead91fe7        97.8K       14 1.17.0      /Users/lucain/.cache/huggingface/hub/datasets--glue/snapshots/f021ae41c879fcabcf823648ec685e3fead91fe7                       
google/fleurs               dataset   129b6e96cf1967cd5d2b9b6aec75ce6cce7c89e8        25.4K        3 refs/pr/1   /Users/lucain/.cache/huggingface/hub/datasets--google--fleurs/snapshots/129b6e96cf1967cd5d2b9b6aec75ce6cce7c89e8             
google/fleurs               dataset   24f85a01eb955224ca3946e70050869c56446805        64.9M        4 main        /Users/lucain/.cache/huggingface/hub/datasets--google--fleurs/snapshots/24f85a01eb955224ca3946e70050869c56446805             
Jean-Baptiste/camembert-ner model     dbec8489a1c44ecad9da8a9185115bccabd799fe       441.0M        7 main        /Users/lucain/.cache/huggingface/hub/models--Jean-Baptiste--camembert-ner/snapshots/dbec8489a1c44ecad9da8a9185115bccabd799fe 
bert-base-cased             model     378aa1bda6387fd00e824948ebe3488630ad8565         1.5G        9             /Users/lucain/.cache/huggingface/hub/models--bert-base-cased/snapshots/378aa1bda6387fd00e824948ebe3488630ad8565              
bert-base-cased             model     a8d257ba9925ef39f3036bfc338acf5283c512d9         1.4G        9 main        /Users/lucain/.cache/huggingface/hub/models--bert-base-cased/snapshots/a8d257ba9925ef39f3036bfc338acf5283c512d9              
t5-base                     model     23aa4f41cb7c08d4b05c8f327b22bfa0eb8c7ad9        10.1K        3 main        /Users/lucain/.cache/huggingface/hub/models--t5-base/snapshots/23aa4f41cb7c08d4b05c8f327b22bfa0eb8c7ad9                      
t5-small                    model     98ffebbb27340ec1b1abd7c45da12c253ee1882a       726.2M        6 refs/pr/1   /Users/lucain/.cache/huggingface/hub/models--t5-small/snapshots/98ffebbb27340ec1b1abd7c45da12c253ee1882a                     
t5-small                    model     d0a119eedb3718e34c648e594394474cf95e0617       485.8M        6             /Users/lucain/.cache/huggingface/hub/models--t5-small/snapshots/d0a119eedb3718e34c648e594394474cf95e0617                     
t5-small                    model     d78aea13fa7ecd06c29e3e46195d6341255065d5       970.7M        9 main        /Users/lucain/.cache/huggingface/hub/models--t5-small/snapshots/d78aea13fa7ecd06c29e3e46195d6341255065d5                     

Done in 0.0s. Scanned 6 repo(s) for a total of 3.4G.
Got 1 error(s) while scanning. Use -vvv to print details.
```
```
➜ huggingface-cli scan-cache -vvv
REPO ID                     REPO TYPE REVISION                                 SIZE ON DISK NB FILES REFS        LOCAL PATH                                                                                                                   
--------------------------- --------- ---------------------------------------- ------------ -------- ----------- ---------------------------------------------------------------------------------------------------------------------------- 
glue                        dataset   9338f7b671827df886678df2bdd7cc7b4f36dffd        97.7K       14 main, 2.4.0 /Users/lucain/.cache/huggingface/hub/datasets--glue/snapshots/9338f7b671827df886678df2bdd7cc7b4f36dffd                       
glue                        dataset   f021ae41c879fcabcf823648ec685e3fead91fe7        97.8K       14 1.17.0      /Users/lucain/.cache/huggingface/hub/datasets--glue/snapshots/f021ae41c879fcabcf823648ec685e3fead91fe7                       
google/fleurs               dataset   129b6e96cf1967cd5d2b9b6aec75ce6cce7c89e8        25.4K        3 refs/pr/1   /Users/lucain/.cache/huggingface/hub/datasets--google--fleurs/snapshots/129b6e96cf1967cd5d2b9b6aec75ce6cce7c89e8             
google/fleurs               dataset   24f85a01eb955224ca3946e70050869c56446805        64.9M        4 main        /Users/lucain/.cache/huggingface/hub/datasets--google--fleurs/snapshots/24f85a01eb955224ca3946e70050869c56446805             
Jean-Baptiste/camembert-ner model     dbec8489a1c44ecad9da8a9185115bccabd799fe       441.0M        7 main        /Users/lucain/.cache/huggingface/hub/models--Jean-Baptiste--camembert-ner/snapshots/dbec8489a1c44ecad9da8a9185115bccabd799fe 
bert-base-cased             model     378aa1bda6387fd00e824948ebe3488630ad8565         1.5G        9             /Users/lucain/.cache/huggingface/hub/models--bert-base-cased/snapshots/378aa1bda6387fd00e824948ebe3488630ad8565              
bert-base-cased             model     a8d257ba9925ef39f3036bfc338acf5283c512d9         1.4G        9 main        /Users/lucain/.cache/huggingface/hub/models--bert-base-cased/snapshots/a8d257ba9925ef39f3036bfc338acf5283c512d9              
t5-base                     model     23aa4f41cb7c08d4b05c8f327b22bfa0eb8c7ad9        10.1K        3 main        /Users/lucain/.cache/huggingface/hub/models--t5-base/snapshots/23aa4f41cb7c08d4b05c8f327b22bfa0eb8c7ad9                      
t5-small                    model     98ffebbb27340ec1b1abd7c45da12c253ee1882a       726.2M        6 refs/pr/1   /Users/lucain/.cache/huggingface/hub/models--t5-small/snapshots/98ffebbb27340ec1b1abd7c45da12c253ee1882a                     
t5-small                    model     d0a119eedb3718e34c648e594394474cf95e0617       485.8M        6             /Users/lucain/.cache/huggingface/hub/models--t5-small/snapshots/d0a119eedb3718e34c648e594394474cf95e0617                     
t5-small                    model     d78aea13fa7ecd06c29e3e46195d6341255065d5       970.7M        9 main        /Users/lucain/.cache/huggingface/hub/models--t5-small/snapshots/d78aea13fa7ecd06c29e3e46195d6341255065d5                     

Done in 0.0s. Scanned 6 repo(s) for a total of 3.4G.
Got 1 error(s) while scanning.
Snapshots dir doesn't exist in cached repo: /Users/lucain/.cache/huggingface/hub/models--tests--fixtures--working_repo_2--FROM_PRETRAINED/snapshots
```

julien-c commented on src/huggingface_hub/commands/_cli_utils.py:
```diff
@@ -0,0 +1,59 @@
+    lines.append(row_format.format(*["-" * w for w in col_widths]))
+    for row in rows:
+        lines.append(row_format.format(*row))
+    return "\n".join(lines)
```
did you change the implem or was it just moved? (asking because we have the same implem in JS in moon-landing)

Wauplin replied:
An exact copy-paste since I didn't want to break anything I didn't know about.

julien-c commented on docs/source/how-to-cache.mdx:
```diff
@@ -0,0 +1,238 @@
+# Manage `huggingface_hub` cache-system
```
this doc is very cool!

julien-c reviewed (approved):
Seems pretty solid to me, approving, but please get another reviewer's feedback before merge?

osanseviero reviewed (approved):
Very cool and excellent docs! Thanks for this :fire:

Wauplin commented on src/huggingface_hub/utils/_cache_manager.py:
```diff
@@ -0,0 +1,446 @@
+
+    repos: Set[CachedRepoInfo] = set()
+    errors: List[CorruptedCacheException] = []
+    for repo_path in Path(cache_dir).resolve().iterdir():
```
Very good catch indeed :+1:

Wauplin replied:
Addressed in https://github.com/huggingface/huggingface_hub/pull/990/commits/f86a7e35ebb9e82f8d161fa191bf41d724289667

Wauplin commented on docs/source/how-to-cache.mdx:
```diff
@@ -0,0 +1,238 @@
+t5-small                    model           970.7M       11 refs/pr/1, main     /Users/lucain/.cache/huggingface/hub/models--t5-small                     
+
+Done in 0.0s. Scanned 6 repo(s) for a total of 3.4G.
+Got 1 error(s) while scanning. Use -vvv to print details.
```
I've used `-vvv` to skip the `-vv` in case we need an extra depth in the future (since I don't know how this tool will evolve).
And also because it doesn't matter so much in my opinion. I'm fine to switch to `-vv` if you find it cleaner though

Wauplin commented on docs/source/how-to-cache.mdx:
````diff
@@ -0,0 +1,238 @@
+        ...
+    ],
+)
+```
````
Yes, agree with you on the need for a utility to delete cached repo/cached revisions. I started to mention this in https://github.com/huggingface/huggingface_hub/pull/990#issuecomment-1226902861 but my plan is definitely to add such a tool. Ideally I would
- merge this PR
- ask on slack if people want to try it before next release (it's already tested but I'd prefer more feedback just because users can have a cache corrupted in a way I did not thought about)
- add the deletion tool and update documentation accordingly
- do all of this before next release so that we ship both the scan and deletion at the same time

About adding an example using `rm` I'm a bit skeptical because revisions have to be carefully deleted (blobs should be deleted only if not referred anymore). Once we have a deletion tool, we can make it work with grep and pipes I think.

Wauplin commented:
Thanks @LysandreJik and @osanseviero for your review and feedback!
There are still two small comments left from Lysandre but apart from that I think we are good to merge.
